Add limits.h inclusion to ssl_tls.c and udp_proxy.c

redtangent opened the pull request:
## Description
The standard C library header limits.h was only being included by `check_config.h` so was being missing if users provided their own config.h that didn't include `check_config.h`.

This fixes #1803.

## Status
**READY**

## Requires Backporting
Yes
Which branch?
[edited by mpg] `mbedtls-2.16` and `mbedtls-2.7`
[quoting redtangent] I've provided backports with PR #2445 and PR #2446 to `mbedtls-2.16` and `mbedtls-2.7`.

## Todos
- [x] Tests
- [x] Documentation
- [x] Changelog updated
- [x] Backported

sbutcher-arm commented:
Let's hold this PR until we have a test case for a custom `config.h` without a `check_config.h`.

hanno-arm commented:
I think this PR should be merged and not held back for the purpose of adding a test case of a custom `config.h` which does not include `check_config.h`. In fact, I think we should rather ensure that, even for custom configs, `check_config.h` is included somewhere, to prevent users from accidentally using malformed configurations. But even if we did that and `check_config.h` was always included somewhere, it's still good style to headers where they are needed, as this PR does.

RonEld commented:
@hanno-arm I agree that it is good practice to add headers when you need them, howeer, it will be good to have a way to test it, and this is why a custom config *without* a call to check_config may be in use

hanno-arm commented:
@RonEld I disagree - I think `check_config.h` should never be omitted, and if that's something we can agree upon (point of discussion, of course), then we shouldn't include a test case exercising the opposite.

RonEld commented:
I agree that `check_config.h` should never be omitted. I just think that we should have a way to test such cases where there are missing header includes.
question: Do we really need `limits.h` included in `check_config.h`?

RonEld commented:
>Question: Do we really need `limits.h` included in `check_config.h`?

Answer: only in one part:
~~~
#if CHAR_BIT != 8
#error "mbed TLS requires a platform with 8-bit chars"
#endif
~~~

I suggest moving this check to `platform.h` ( or the c file) , where it is more suited, and remove the `limits.h` inclusion in `check_config.h`

redtangent commented:
Added the test to `all.sh` as requested. I toyed with the idea of moving `limits.h` out of `check_config.h` but then the problem will become wherever it goes, every single include of `config.h` and therefore `check_config.h` will also have to include. Better keep it where it is in my view.

RonEld commented:
@redtangent Thank you for adding the test!
>but then the problem will become wherever it goes

I don't understand your point. This is what we want to check. If a file that will need `limits.h` will fail to compile now,  it wll need to add the `limits.h` file. In my brief check, I removed the `limits.h` from `check_config.h`, included it to wherever missing( ssl_tls.c and udp_proxy.c) and didn't encounter compilation errors. Adding to `platform.h` the check for 
~~~
#if CHAR_BIT != 8
#error "mbed TLS requires a platform with 8-bit chars"
#endif
~~~
should not add a problem, IMHO.
I agree with @hanno-arm that we should be even stricter, byenforcing adding the `check_config.h` also to custom configs

sbutcher-arm commented:
If `limits.h` is not included, `CHAR_BIT`will be undefined and evaluate to 0. If we don't include `limits.h` locally alongside the use of `CHAR_BIT`, but instead in `platform.h`, then surely, we need to include `platform.h` everywhere that `check_config.h` or `config.h` is included.

RonEld commented:
But the check for `CHAR_BIT` as commented is:
~~~
/*
 * We assume CHAR_BIT is 8 in many places. In practice, this is true on our
 * target platforms, so not an issue, but let's just be extra sure.
 */
~~~
So, the need for `limits.h` in `check_config.h` is *only* for using `CHAR_BIT`
If we move the inclusion of `limits.h` *and* the `CHAR_BIT` check to `platform.h`, we will still have the check, and we don't need to include `platform.h` everywhere `check_config.h` or `config.h` are included.
`platform.h` is already included in most of the files ( if not all ), so we will encounter the `CHAR_BIT` check at least once.
If we *must* keep the check for `CHAR_BIT` inside `check_config.h` then I agree we need to keep `limits.h` as well, but I am not convinced yet that this is a must.

RonEld commented:
A quick grep for `platfortm.h` shows the following:
~~~
library/ssl_tls.c:#include "mbedtls/platform.h"
library/x509_crt.c:#include "mbedtls/platform.h"
library/bignum.c:#include "mbedtls/platform.h"
library/net_sockets.c:#include "mbedtls/platform.h"
library/pkwrite.c:#include "mbedtls/platform.h"
library/cipher.c:#include "mbedtls/platform.h"
library/x509_crl.c:#include "mbedtls/platform.h"
library/asn1write.c:#include "mbedtls/platform.h"
library/ctr_drbg.c:#include "mbedtls/platform.h"
library/aria.c:#include "mbedtls/platform.h"
library/pk_wrap.c:#include "mbedtls/platform.h"
library/x509.c:#include "mbedtls/platform.h"
library/ccm.c:#include "mbedtls/platform.h"
library/des.c:#include "mbedtls/platform.h"
library/md.c:#include "mbedtls/platform.h"
library/poly1305.c:#include "mbedtls/platform.h"
library/timing.c:#include "mbedtls/platform.h"
library/sha1.c:#include "mbedtls/platform.h"
library/pkcs11.c:#include "mbedtls/platform.h"
library/arc4.c:#include "mbedtls/platform.h"
library/ecjpake.c:#include "mbedtls/platform.h"
library/entropy.c:#include "mbedtls/platform.h"
library/entropy.c:#include "mbedtls/platform.h"
library/ssl_srv.c.orig:#include "mbedtls/platform.h"
library/sha256.c:#include "mbedtls/platform.h"
library/platform.c:#include "mbedtls/platform.h"
library/md_wrap.c:#include "mbedtls/platform.h"
library/aes.c:#include "mbedtls/platform.h"
library/hmac_drbg.c:#include "mbedtls/platform.h"
library/dhm.c:#include "mbedtls/platform.h"
library/chacha20.c:#include "mbedtls/platform.h"
library/gcm.c:#include "mbedtls/platform.h"
library/ssl_cookie.c:#include "mbedtls/platform.h"
library/memory_buffer_alloc.c:#include "mbedtls/platform.h"
library/ecp.c:#include "mbedtls/platform.h"
library/ssl_ticket.c:#include "mbedtls/platform.h"
library/rsa.c:#include "mbedtls/platform.h"
library/md2.c:#include "mbedtls/platform.h"
library/asn1parse.c:#include "mbedtls/platform.h"
library/cmac.c:#include "mbedtls/platform.h"
library/pem.c:#include "mbedtls/platform.h"
library/ssl_ciphersuites.c:#include "mbedtls/platform.h"
library/nist_kw.c:#include "mbedtls/platform.h"
library/x509_csr.c:#include "mbedtls/platform.h"
library/entropy_poll.c:#include "mbedtls/platform.h"
library/pkparse.c:#include "mbedtls/platform.h"
library/ripemd160.c:#include "mbedtls/platform.h"
library/ssl_cache.c:#include "mbedtls/platform.h"
library/ssl_cli.c:#include "mbedtls/platform.h"
library/ssl_srv.c:#include "mbedtls/platform.h"
library/base64.c:#include "mbedtls/platform.h"
library/oid.c:#include "mbedtls/platform.h"
library/debug.c:#include "mbedtls/platform.h"
library/pkcs5.c:#include "mbedtls/platform.h"
library/md5.c:#include "mbedtls/platform.h"
library/chachapoly.c:#include "mbedtls/platform.h"
library/md4.c:#include "mbedtls/platform.h"
library/camellia.c:#include "mbedtls/platform.h"
library/sha512.c:#include "mbedtls/platform.h"
library/error.c:#include "mbedtls/platform.h"
library/error.c:#include "mbedtls/platform.h"
library/xtea.c:#include "mbedtls/platform.h"
library/cipher_wrap.c:#include "mbedtls/platform.h"
~~~
So, I believe this file will be included at least once by any application.
However, I understand that it might clash with the modular design of the library, so I won't insist on this suggestion.

redtangent commented:
I think in practice, @RonEld, you're right, everything includes `platform.h`, but that's not intentional nor part of the philosophy of the library, which aims to be very modular.

So, best, in my view, to not place the check there, but to keep it in `check_config.h` which should be included everywhere.

Maybe we should move including `check_config.h` outside of every `config.h`, which we now know some people omit, and instead include it in every `.c` file? That would remove the need for the `all.sh` check in this PR. 

I personally think that's a better fix. Let me know what you think.

RonEld commented:
>Maybe we should move including check_config.h outside of every config.h, which we now know some people omit, and instead include it in every .c file?

Yes, I agree with you it may be a better solution. I think that in the long run, we should have one common header file, which within will include the `config.h`, `check_config.h` and perhaps some additional common information, such as error codes, but this is out of the scope of this PR.
Let's go with your suggestion, by removing from `config,h` and including it in every `.c` file.

sbutcher-arm commented:
Thanks @RonEld. To move this along, let's agree that `check_config.h` should be included in header module regardless, and we can drop the tests in `all.sh` as unnecessary.

redtangent commented:
I've added inclusion of `check_config.h` into all library files, so they'll catch issues even if the user provides their own `config.h` file.

I've also removed the `all.sh` test, as this isn't necessary if `check_config.h` is included in all files.

redtangent commented:
@RonEld / @hanno-arm - Could you please re-review my PR?

RonEld commented:
Although the '.c' will be compiled and used, I think it is better to add the `check_config.h` file in the header files, as these include the configuration files( some of them ) there is a Pr #1375 that adds inclusions to all headers that need configuration file.

sbutcher-arm commented:
@hanno-arm - Could you please re-review so we can get this merged? Thanks.

mpg commented:
@redtangent This is nearly ready for merge, except backports are missing. Do you want to provide backport or should we take care of that?

redtangent commented:
I've provided backports with PR #2445 and PR #2446 to `mbedtls-2.16` and `mbedtls-2.7`.

gilles-peskine-arm commented:
Unfortunately, after all this time, the PR has accumulated merge conflicts. And a few more source files have appeared and need their include directives updated.

gilles-peskine-arm commented:
I'm not going to veto this PR, but I don't like the chosen approach which requires changing every source file. If we're going to do that, let's make every library source file include `common.h` (and yes, I have a PR in flight that creates it but doesn't add the `#include`s: https://github.com/ARMmbed/mbedtls/pull/3121).

sbutcher-arm commented:
Rebased, fixed the conflicts, and added the missing files and also `<limits.h>` to those files that were missing it. 

@RonEld / @hanno-arm - would you mind re-reviewing this? If you're not available, just say and I'm sure we can find a substitute reviewer.

sbutcher-arm commented:
CI is failing because of some issue in PSA crypto, unrelated to this PR. I'll rebase, and hopefully it will all go away.

mpg commented:
Unfortunately there are still CI failures, some of them related to `check_config.h`.

sbutcher-arm commented:
This PR is now passing all CI except for Mbed OS. Mbed OS is failing because the absence of `#include "check_config.h"` breaks one of the Mbed OS import scripts. I've fixed that with [PR #12998](https://github.com/ARMmbed/mbed-os/pull/12998).

To avoid this PR becoming yet again quickly outdated and requiring another rebase, my recommendation would be to review and merge without waiting for that PR to be merged - if possible.

Otherwise this PR is now ready for review again.

sbutcher-arm commented:
@dgreen-arm / @hanno-arm - Would you be kind enough to review this old PR and it's siter backports, as your names are already on them? If you do - I *promise* a review in return, for the PR of your choice, as big or as small as you want.

gilles-peskine-arm commented on library/ssl_tls.c:
```diff
@@ -50,6 +50,7 @@
 #include "mbedtls/platform_util.h"
 #include "mbedtls/version.h"
 
+#include <limits.h>
```
`ssl_tls.c` no longer uses anything from `limits.h`: now only `ssl_msg.h` does. This is not a blocker: including a standard header which doesn't get used is harmless.

However, 4 other files need `limits.h` on this branch: `bignum.c`, `ctr_drbg.c`, `psa_crypto.c` and `x509_crt.c`. I checked and `development` doesn't need any other as of today. Please add those, otherwise it defeats the purpose of the PR. And please re-check test suites and sample programs.

gilles-peskine-arm commented on ChangeLog:
```diff
@@ -37,6 +37,9 @@ Bugfix
      a warning with some compilers. Fix contributed by irwir in #2856.
    * Fix a function name in a debug message. Contributed by Ercan Ozturk in
      #3013.
+   * Add missing limits.h standard C library header to ssl_tls.c and udp_proxy.c
+     which was only including it via check_config.h previously, which may not be
+     included in custom configuration files. Fixes #1803
```
This changelog entry would get added to the section for the already-released version 2.22.0. We no longer merge changelog entries manually. Please create a file in `ChangeLog.d` instead. This doesn't require a rebase.

gilles-peskine-arm commented on 3rdparty/everest/library/Hacl_Curve25519_joined.c:
```diff
@@ -24,6 +24,7 @@
 #else
 #include MBEDTLS_CONFIG_FILE
 #endif
+#include "mbedtls/check_config.h"
```
[As I've stated before](https://github.com/ARMmbed/mbedtls/pull/1999#issuecomment-615460281), I don't like this approach. Every library source file is changed to include something that doesn't really concern it.

Furthermore this is unrelated to the original objective of the pull request. It neither fixes the problem nor contributes to making the problem less likely to happen again.

I wouldn't have vetoed the PR on this basis, but since the PR needs rework anyway, please remove the inclusion of `check_config.h` everywhere.

To include `check_config.h` everywhere, on `development`, add it to `library/common.h`, which [was added to `development` today](https://github.com/ARMmbed/mbedtls/pull/3121) (and we'll also need to include `common.h` everywhere, which isn't done yet).

sbutcher-arm commented:
Thanks for the review @gilles-peskine-arm. I guess you weren't aware, but the points you are making I was already aware of and are exactly why I told you the PR needed rework - and those changes have only arisen out of changes to `development` while this PR was waiting for reviews - which is why I was expressing frustration. 

Ideally if this PR had been reviewed more quickly, then it would not have been effected by `common.h` nor the ChangeLog changes, nor the changes to which file needed the change in the first place.

gilles-peskine-arm commented:
@sbutcher-arm Do you intend to rework this PR soon? I'm working on another change (moving stuff to the newly created `common.h`) which runs into the `limits.h` inclusion problem and therefore conflicts with this PR. If you intend to rework soon, I'll wait until this is merged. Otherwise I'll make my PR which will fix this bug in development, and I can either take over the backports or leave them for later.

sbutcher-arm commented:
@gilles-peskine-arm - After last week I talked to @danh-arm and passed all my open _personal_ PR's including this one over to the core team. I am no longer doing any work on Mbed TLS in my free time and am pursuing my interests in crypto elsewhere.

Any future PR's will be strictly business related.

gilles-peskine-arm commented on include/mbedtls/config.h:
```diff
@@ -3688,6 +3688,4 @@
 #include MBEDTLS_USER_CONFIG_FILE
 #endif
 
-#include "mbedtls/check_config.h"
```
[There's one place where `check_config.h` _modifies_ the configuration](https://github.com/ARMmbed/mbedtls/blob/mbedtls-2.22.0/include/mbedtls/check_config.h#L47).

```
#if defined(_WIN32)
#if !defined(MBEDTLS_PLATFORM_C)
#error "MBEDTLS_PLATFORM_C is required on Windows"
#endif

/* Fix the config here. Not convenient to put an #ifdef _WIN32 in config.h as
 * it would confuse config.py. */
#if !defined(MBEDTLS_PLATFORM_SNPRINTF_ALT) && \
    !defined(MBEDTLS_PLATFORM_SNPRINTF_MACRO)
#define MBEDTLS_PLATFORM_SNPRINTF_ALT
#endif

#if !defined(MBEDTLS_PLATFORM_VSNPRINTF_ALT) && \
    !defined(MBEDTLS_PLATFORM_VSNPRINTF_MACRO)
#define MBEDTLS_PLATFORM_VSNPRINTF_ALT
#endif
#endif /* _WIN32 */
```

If `check_config.h` is no longer included from `config.h`, and thus only included from library code and not from application code, this won't do. Application code and library code must have the same configuration.

gilles-peskine-arm replied:
The only code in the library that is functionally affected by `MBEDTLS_PLATFORM_SNPRINTF_ALT` or `MBEDTLS_PLATFORM_VSNPRINTF_ALT` is in `platform.h` and `platform.c`. Application code that is affected by these macros can be reasonably expected to include `mbedtls/platform.h`. Therefore I think this hack can be adapted to fit in `platform.h` instead.

Feature detection (`version_features.c`, `query_config`) would no longer report these alt symbols correctly. I think that an acceptable change would be to not set the symbols, but cause their effect in `platform.h` regardless.

gilles-peskine-arm replied:
Fortunately, it turns out that changing the configuration is not necessary: `platform.h` will do its thing (activate an snprintf wrapper) regardless. https://github.com/ARMmbed/mbedtls/pull/3453

gilles-peskine-arm commented:
I'm working on a PR that currently includes the modifications made here: fixing #1803 and not including `check_config.h` from `config.h`. But I'm still working out the implications of the second part. If I can't make it work, I'll make a separate PR that fixes #1803 and nothing else.

daverodgman commented:
This issue is tracked in #1803 - closing this PR as probably this needs a different approach.
